### DISCUSSION TITLE
How to trigger a full window redraw? (Similar to when user manual resize it)

### DISCUSSION THREAD

**Discussion by sn4k3** [AvaloniaUI/Avalonia#8082] [Q&A]
I need to trigger a full window redraw, similar to what happens when user manual resize the window by the borders.
Reason: Avalonia window AutoSize with MaxWidth/Height is bugged since ever (elements will strech to infinity never respecting the set window MaxSize) unless user manual resize the window so control will constrain correctly. I open some issues but no fix: https://github.com/AvaloniaUI/Avalonia/issues/6441 https://github.com/AvaloniaUI/Avalonia/issues/5466 https://github.com/AvaloniaUI/Avalonia/issues/4789
In a UI framework a bug like this is kinda sloppy. This should take serious review and proper fix! I understand that a autosize will strech to infinity up to the content but when MaxWidth/Height is set it should be respected no matter of the mode, it already do when user resize window, why not doing it on constructor/render too? I know this is caused by `Window.Width and Height = NaN` but it should be looking at bounds and compare to MaxWidth/Height before draw the elements.

So to be able to deliver a fix on my app i do many hacks like, set to manual, resize window, set position, set again to autosize, etc. Which is a pain and not optimal.
However if user resize window by by px amout it will fix all the view! So how can i trigger a full redraw of window to produce same effect? I tried all `Invalidate*` methods, none works to produce the same effect as a manual resize would do.

Window.MaxWidth and MaxHeight is factorized to `screen resolution / 2`, produces:

![UVtools_2022-05-05_14-10-44](https://user-images.githubusercontent.com/113281/166930035-35f57402-375d-4680-be1e-b68e1340c85a.png)

Now if user resize window by any value it fix the problem:
![UVtools_2022-05-05_14-11-44](https://user-images.githubusercontent.com/113281/166930135-37600cc6-7a78-4221-94b5-17e0f4f5a767.png)

This is what i want to trigger by code on `OnLoaded`

Note, the following code doesn't work for the same porpouse (Not equal to user manual resize) and content will still be overflowing:

```C#
Width = 10;
Height = 10;
Width = 600;
Height = 500;
```

The downside of doing a resize to fix this problem is that AutoSize mode is lost, since now window will have fixed size and no longer adjust to content. And windows startup position will be invalidated too if we change the window size. That's why this need a proper fix.

**Comment by timunie** [comment]
Hi ,

Do you think you are able to send a PR which fixes your issue? Maybe this can help to speed up things. I can't promise as I am not a core member, but I got my issues fixed that way. 

As a workaround, consider to set MaxHeight to your ScrollViewer. Or the root Grid. 

Happy coding
Tim

**Reply by sn4k3** [reply]
I would do, but i dont know where to start, i never dig into avalonia sources. And i guess this is not a "easy fix".
However i have a guess. Since windows Width and Height is NaN controls are free to strech whatever they need, the problem happens on that condition. So instead of windows Width/Height, it should look at Bounds which always have a size, and if greater than set MaxWidth/Height constrain bounds to MaxWidth/Height. I have no clue how to do that properly on avalonia core.

> As a workaround, consider to set MaxHeight to your ScrollViewer. Or the root Grid.

Same happens, without a fixed width/height problem will show

**Reply by timunie** [reply]
Hi

I think the issue must be somewhere here: https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.Controls/Window.cs 

Good luck 👍 
Tim

**Reply by sn4k3** [reply]
Thank you, i was able to fix and was much easier than i tought!

**Comment by Gillibald** [comment]
Do you have some code to share? This might help a lot.

**Reply by timunie** [reply]
@Gillibald see the issues linked above. I tried the xaml from the first issue linked above and I was able reproduce the issue (I used the Sandbox App for testing)

**Comment by sn4k3** [comment]
Finally, i was able to fix the problem. It was so simple that i wonder why this bug was here since ever and never got any serious attention...

PR: https://github.com/AvaloniaUI/Avalonia/pull/8093
